### PR TITLE
feat: Deny install/upgrade of packages which server dep does not meet the current Appium version

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
## Proposed changes

We already have a similar check for locally installed extensions (that one only prints warnings). Although it does not seem logical to allow modules installation if these are not compatible.

The check is only applied to packages managed by NPM

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

### CODE REVIEW

File: packages/appium/lib/cli/extension-command.js [hunk 1]
```diff
@@ -225,6 +240,29 @@ class ExtensionCliCommand {
...
+      return;
+    }
+
+    await spinWith(this.isJsonOutput, `Checking if '${pkgName}' is compatible`, async () => {
+      const [serverVersion, extVersionRequirement] = await getRemoteExtensionVersionReq(installSpec);
+      if (serverVersion && extVersionRequirement && !semver.satisfies(serverVersion, extVersionRequirement)) {
+        throw this._createFatalError(
+          `'${installSpec}' cannot be installed because the server version it requires (${extVersionRequirement}) ` +
+          `does not meet the currently installed one (${serverVersion}). Please install ` +
+          `a compatible server version first.`
```
Comment (eglitise) [inline packages/appium/lib/cli/extension-command.js:260]:
```suggestion
          `does not meet the currently installed one (${serverVersion}). Please install ` +
          `a compatible server version, or use a different version of '${pkgName}'.`
```

**Review by eglitise — APPROVED** [review]
Just to confirm, all official extensions have `"peerDependencies": { "appium": "^2.0.0" }` (or some earlier beta version), correct? So their installs should not be affected yet. It's only if I have e.g. Appium `2.0.0` and try to install an extension with `^2.2.0` or something, that an error would be raised?

Also I wonder if it's possible to add a CI test for this?

File: packages/appium/lib/cli/extension-command.js (hunk 1 above)
Comment (mykola-mokhnach) [reply packages/appium/lib/cli/extension-command.js:260]: > or use a different version of '${pkgName}'

This message is also shown on upgrade, so I was not sure if it wouldn't be confusing

**Comment by mykola-mokhnach** [comment]
> Just to confirm, all official extensions have "peerDependencies": { "appium": "^2.0.0" } (or some earlier beta version), correct? So their installs should not be affected yet. It's only if I have e.g. Appium 2.0.0 and try to install an extension with ^2.2.0 or something, that an error would be raised?

Yes, that's correct. This feature is more relevant for cases where one is running an older Appium server, which does not have all the features that a newer extension version requires. I don't think driver or extension authors have any reasons to cap the minor or patch version of the server by default

**Comment by mykola-mokhnach** [comment]
> Also I wonder if it's possible to add a CI test for this?

For that we would have to have some obsolete non matching extension available in NPM 🤷

File: packages/appium/lib/cli/extension-command.js (hunk 1 above)
Comment (mykola-mokhnach) [reply packages/appium/lib/cli/extension-command.js:260]: also, it is not clear if there is `a different version` of the package, which would be compatible to the current server version

**Review by KazuCocoa — APPROVED** [review]
lgtm

**Comment by KazuCocoa** [comment]
Should we update `peerDependencies` to `@2.0.0` which still have `-beta.35`?
https://github.com/search?q=repo%3Aappium%2Fappium%20peerDependencies&type=code
(let me check semver.satisfies..)

**Comment by KazuCocoa** [comment]
nvm, seems good:
```
> s.satisfies('2.0.0', '^2.0.0-beta.35')
true
```

**Pull request merged**